File: Practica 3/CivitasRuby/pkg/RubyApplication1-0.0.1/lib/sorpresa.rb
#encoding:utf-8
  
require_relative 'diario.rb'
require_relative 'tipo_sorpresa.rb'
require_relative 'mazo_sorpresas.rb'
require_relative 'tablero.rb'
require_relative 'jugador.rb'
require_relative 'casilla.rb'

module Civitas
  
class Sorpresa

# @brief Constructor de la clase.
# @param tipo, el tipo de sorpresa
# @param texto, el nombre de la sorpresa
# @return Un objeto de clase sorpresa  
def initialize(texto, tipo)
  init()
  @tipo = tipo
  @texto = texto
end
  
# @brief Método que hace de Constructor de la clase. Resto de sorpresas
# @param tipo, el tipo de sorpresa
# @return Un objeto de clase sorpresa  
def self.new_valor(tipo, valor, texto)
  @valor = valor
  new(texto, tipo)
end

# @brief Método que hace de Constructor de la clase. Evitar la cárcel
# @param tipo, el tipo de sorpresa
# @param mazo, el mazo con en el que se juega
# @return Un objeto de clase sorpresa  
def self.new_mazo(tipo, mazo)
  @mazo = mazo
  new("MAZO", tipo)
end

# @brief Método que hace de Constructor de la clase. Enviar a Cárcel
# @param tipo, el tipo de sorpresa
# @param tablero, el tablero en el que juega
# @return Un objeto de clase sorpresa  
def self.new_tablero(tipo, tablero)
  @tablero = tablero
  new("A la cárcel!", tipo)
end

# @brief Método que hace de Constructor de la clase. Enviar jugador a otra cas
# @param tipo, el tipo de sorpresa
# @param tablero, el tablero en el que juega
# @param valor, el valor de la sorpresa
# @param texto, el nombre de la sorpresa
# @return Un objeto de clase sorpresa
def self.new_tablero_valor(tipo, tablero, valor, texto)
  @valor = valor
  @tablero = tablero
  new(texto, tipo)
end

####################################################################
## Resto de métodos
####################################################################

# @brief Aplica una sorpresa en funcion del autrbuto de instancia @tipo
# @param actual, el indice del jugador
# @param todos, el vector de jugadores
# @return Es void
public
def aplicar_a_jugador(actual, todos)
    case @tipo
      when TipoSorpresa::IRCARCEL
        aplicar_a_jugador_ir_carcel(actual,todos)
      when TipoSorpresa::IRCASILLA
        aplicar_a_jugador_ir_a_casilla(actual,todos)
      when TipoSorpresa::SALIRCARCEL
        aplicar_a_jugador_salir_carcel(actual,todos)
      when TipoSorpresa::PAGARCOBRAR
        aplicar_a_jugador_pagar_cobrar(actual,todos)
      when TipoSorpresa::PORCASAHOTEL
        aplicar_a_jugador_por_casa_hotel(actual,todos)
      when TipoSorpresa::PORJUGADOR
        aplicar_a_jugador_por_jugador(actual,todos)
    end
end

# @brief Modifica la posición del jugador, actualizándose la casilla donde está
# @param actual, el indice del jugador
# @param todos, el vector de jugadores
# @return Es void
private
def aplicar_a_jugador_ir_a_casilla(actual,todos)
  if(jugador_correcto(actual,todos))
    informe(actual,todos)
    casilla_actual = todos[actual].num_casilla_actual
    tirada = @tablero.calcular_tirada(casilla_actual,@tablero.carcel)
    casilla_final = casilla_actual+@valor
    
    if (casilla_final <= 0)
      tirada = tirada +@tablero.size
    end
    
    nueva_posicion = @tablero.nueva_posicion(casilla_actual, tirada)
    todos.at(actual).mover_a_casilla(nueva_posicion)
    @tablero.get_casilla(nueva_posicion).recibe_jugador(actual, todos)
  end
end

# @brief Aplica la sorpresa de encarcelar usando el método de jugador encarcelar
# @param actual, el indice del jugador
# @param todos, el vector de jugadores
# @return Es void
private
def aplicar_a_jugador_ir_carcel(actual, todos)
    if(jugador_correcto(actual,todos))
      informe(actual,todos)
      todos[actual].encarcelar(@tablero.num_casilla_carcel)
    end
end

# @brief Aplica la sorpresa de modificar el saldoe enfuncion de @valor
# @param actual, el indice del jugador
# @param todos, el vector de jugadores
# @return Es void
private
def aplicar_a_jugador_pagar_cobrar(actual,todos)
  if(jugador_correcto(actual,todos))
    informe(actual,todos)
    todos[actual].modificar_saldo(@valor)
  end
end

# @brief Aplica la sorpresa de modificar el saldo multiplicar @valor con la
#   cantidad de casas y hoteles que el jugador posee
# @param actual, el indice del jugador
# @param todos, el vector de jugadores
# @return Es void
private
def aplicar_a_jugador_por_casa_hotel(actual,todos)
  if(jugador_correcto(actual,todos))
    informe(actual,todos)
    todos[actual].modificar_saldo(@valor * todos[actual].cantidad_casas_hoteles())
  end
end

# @brief Llama al método pagar y cobrar y todos los jugadores le dan @valor
#   al jugador actual, y el jugador actual recibe @valor * num_jugadores 
# @param actual, el indice del jugador
# @param todos, el vector de jugadores
# @return Es void
private
def aplicar_a_jugador_por_jugador(actual,todos)
  if(jugador_correcto(actual,todos))
    informe(actual,todos)
    
    a_cobrar = Sorpresa.new_valor(TipoSorpresa::PAGARCOBRAR,(todos.length-1)* @valor, "cobrar")
    a_pagar = Sorpresa.new(TipoSorpresa::PAGARCOBRAR,-1*@valor , "pagar")
    
    for i in (0..todos.length)
      if(i != actual) #El jugador "actual" no paga
        a_pagar.aplicar_a_jugador(i,todos)
      end
    end
   
    a_cobrar.aplicar_a_jugador(actual,todos) #Llama a modifica saldo 
    
  end
end

# @brief Si nadie tiene la sorpresa para evitar la carcel, la obtine el
#   jugador actual
# @param actual, el indice del jugador
# @param todos, el vector de jugadores
# @return Es void  
private
def aplicar_a_jugador_salir_carcel(actual,todos)
  if(jugador_correcto(actual,todos))
    informe(actual,todos)
  end
end

# @brief Informa de una sorpresa al jugador todos.[actual]
# @param actual, el indice del jugador
# @param todos, el vector de jugadores
# @return Es void
# @pre Debe existir el jugador
private
def informe(actual,todos)
  if(jugador_correcto(actual,todos))
    Diario.instance.ocurre_evento("Se ha aplicado una sorpresa a " + todos[actual].nombre)
  end
end

# @brief Inicia los atributos de instancia de valor, tablero y mazo
private
def init()
  @valor = -1
  @tablero = nil
  @mazo = nil
end

# @brief Comprueba si un indice existe en un vector de jugadores
# @param actual, el indice a comprobar
# @param todos, el vector
# @return True si es un indice valido para acceder a los elementos de "todos"
public
def jugador_correcto(actual, todos)
  resultado = false
  
  if(todos.length > actual && actual >= 0) #primera pos del vector
    resultado = true
  end
end

# @brief Si la sorpresa es la que evita la carcél, inhabilita la carta especial 
#   en el mazo de sorpresa
# @return Es void
public
def salir_del_mazo()
  if (@tipo == TipoSorpresa::SALIRCARCEL)
    @mazo.inhabilitar_carta_especial(this)
  end
end

# @brief Devuelve el nombre de la sorpresa(tipo)
# @return Un string con el nombre
public
def to_s()
  "TipoSorpresa: " + @tipo.to_s + "\nValor: " + @valor.to_s + "\nTexto: "+@texto
end

# @brief Si la sorpresa es la que evita la carcél, habilita la carta especial 
#   en el mazo de sorpresa
# @return Es void
public
def usada()
  if (@tipo == TipoSorpresa::SALIRCARCEL)
    @mazo.habilitar_carta_especial(this)
  end
end
    
end # Fin class Sorpresa
end # Fin module Civitas

File: Practica 3/CivitasJava/src/civitasjava/Sorpresa.java

package civitasjava;

import java.util.ArrayList;

public class Sorpresa {
    private String texto;
    private int    valor; // A que casilla te vas, cuanto dinero recibes....
    
    private Tablero       tablero;
    private MazoSorpresas mazo;
    private TipoSorpresa  tipo;
    
    
    /**
     * @brief Aplica una sorpresa en funcion del autrbuto de instancia @tipo
     * @param actual, el indice del jugador
     * @param todos, el vector de jugadores
     * @return void
     */
    void aplicarAJugador(int actual, ArrayList<Jugador> todos){
       switch(this.tipo){
         case IRCASILLA   : aplicarAJugador_irAcasilla  (actual, todos); break;
         case IRCARCEL    : aplicarAJugador_irCarcel    (actual, todos); break;
         case PAGARCOBRAR : aplicarAJugador_pagarCobrar (actual, todos); break;
         case PORCASAHOTEL: aplicarAJugador_porCasaHotel(actual, todos); break;
         case PORJUGADOR  : aplicarAJugador_porJugador  (actual, todos); break;
         case SALIRCARCEL : aplicarAJugador_salirCarcel (actual, todos); break;
      }            
    }
    
    
    /**
     * @brief Modifica la posicion del jugador, actualizáad la casilla donde esta
     * @param actual, el indice del jugador
     * @param todos, el vector de jugadores
     * @return void
     */
    private void aplicarAJugador_irAcasilla  (int actual, ArrayList<Jugador> todos){
        
        if( this.jugadorCorrecto(actual, todos) ){
            
            this.informe(actual, todos); 
            int casillaActual = todos.get(actual).getNumCasillaActual();
            
            // Cuantas posiciones nos faltan hasta llegar a la casilla valor
            int tirada = tablero.calcularTirada(casillaActual, valor);
            
            // Calculamos la posicion donde nos encontramos al desplazarnos
            int nuevaPosicion = tablero.nuevaPosicion(casillaActual, tirada);
            
            // Movemos al jugador a esa nueva posicion
            todos.get(actual).moverACasilla(nuevaPosicion);
            
            // Indicamos a la casilla que reciba al jugador
            tablero.getCasilla(nuevaPosicion).recibeJugador(actual, todos);
            
            // Parece innecesario usar calcularTirada y nuevaPosicion pq
            // sorpresa.valor guarda la casilla donde hay que ir, pero
            // hace falta para guardar si se pasara por la salida (:
        }
    }
    
    
    /** @brief Aplica la sorpresa de encarcelar usando el método de jugador encarcelar
     *  @param actual, el indice del jugador
     *  @param todos, el vector de jugadores
     *  @return void
     */
    private void aplicarAJugador_irCarcel    (int actual, ArrayList<Jugador> todos){
        if( jugadorCorrecto(actual, todos) ){
            this.informe(actual, todos);
            todos.get(actual).encarcelar(this.tablero.getCarcel()); 
        }
    }
        
    
    /**
     * @brief Aplica la sorpresa de modificar el saldo en funcion de @valor
     * @param actual, el indice del jugador
     * @param todos, el vector de jugadores
     * @return void
     */
    private void aplicarAJugador_pagarCobrar (int actual, ArrayList<Jugador> todos){
        
        if( jugadorCorrecto(actual, todos) ){
            
            this.informe(actual, todos);
            todos.get(actual).modificarSaldo(this.valor);
        }
    }
         
    
    /**
     * @brief Modifica el saldo del jugador multiplicando valor con la
     *        cantidad de casas y hoteles que se poseen
     * @param actual, el indice del jugador
     * @param todos, el array de jugadores
     * @return void
     */
    private void aplicarAJugador_porCasaHotel(int actual, ArrayList<Jugador> todos){ 
        
      if( jugadorCorrecto(actual, todos) ){
        this.informe(actual, todos);
        
        Jugador jugador = todos.get(actual);
        jugador.modificarSaldo(this.valor * jugador.cantidadCasasHoteles());
      }
    }
     
    
    /**
     * @brief Todos los jugadores dan el valor de la sorpresa al jugador actual
     * @param actual, el indice del jugador
     * @param todos, el vector de jugadores
     * @return void
     */
    private void aplicarAJugador_porJugador  (int actual, ArrayList<Jugador> todos){
      
      if( jugadorCorrecto(actual, todos) ){
        
        this.informe(actual, todos);
        int numJugadores = todos.size();
        
        // Creamos dos Sorpresas con los valores de saldo a modificar
        Sorpresa restoPaga    = new Sorpresa
                (TipoSorpresa.PAGARCOBRAR, -this.valor                  , "");
        
        Sorpresa actualRecibe = new Sorpresa
                (TipoSorpresa.PAGARCOBRAR,  this.valor*(numJugadores -1), "");
        
        // Cobramos a todos los jugadores menos el que recibe
        for(int i = 0; i < numJugadores; i++){
            if(i != actual)
                restoPaga.aplicarAJugador(i, todos);    
        }
        
        // Ingresamos el dinero al jugador actual
        actualRecibe.aplicarAJugador(actual, todos);
      }
    }
    
    
    // Aqui hay un comentario pregunta SIN COMPROBAR CON ALVARO
    /**
     * @brief Si nadie tiene la sorpresa para evitar la carcel, la obtine el jugador actual
     * @param actual, el indice del jugador
     * @param todos, el vector de jugadores
     * @return void
     */
    private void aplicarAJugador_salirCarcel (int actual, ArrayList<Jugador> todos){
      if( jugadorCorrecto(actual, todos) ){
        this.informe(actual, todos);
        
        // Preguntamos si alguien tiene la sorpresa para evitar la carcel
        boolean tieneSalvoconducto = false;
        for(int i = 0; i < todos.size() && !tieneSalvoconducto; i++)
            tieneSalvoconducto = todos.get(i).tieneSalvoconducto();
        
        // Si nadie la tiene, la obtiene el jugador actual
        if(!tieneSalvoconducto){
            todos.get(actual).obtenerSalvoconducto(this);
            this.salirDelMazo();
        }
        
        // Por que no miramos simplemente si alguien tiene la sorpresa 
        // haciendo mazo.cartasEspeciales.size() > 0 ??
      }
    }
       
    
    /**
     * @brief Informa de una sorpresa al jugador pasado como parametro
     * @param actual, el indice del jugador
     * @param todos, el vector de jugadores
     * @return void
     * @pre Debe existir el jugador
     */
    private void informe(int actual, ArrayList<Jugador> todos){
        Diario.getInstance().ocurreEvento
                                  ( "Se ha aplicado al jugador " 
                                   + todos.get(actual).getNombre()
                                   + " la sorpresa\n"
                                   + this.toString()  
                                  );
    }
    
    
    /**
     * @brief Inicia los atributos de instancia de valor, tablero y mazo
     */
    private void init(){
        this.valor   = -1;
        this.mazo    = null;
        this.tablero = null;
    }
    
    
    /**
     * @brief Comprueba si un indice existe en un vector de jugadores
     * @param actual, el indice a comprobar
     * @param todos, el vector
     * @return TRUE si el indice es valido para acceder a los elementos de @todos
     */
    public boolean jugadorCorrecto(int actual, ArrayList<Jugador> todos){
        return actual >= 0 && todos.size() > actual;
    }
    
    
    /**
     * @brief Si la sorpresa es la que evita la carcel, inhabilita la carta 
     *        especial en el mazo de sorpresa
     * @return void
     */
    void salirDelMazo(){
        if( this.tipo == TipoSorpresa.SALIRCARCEL ) 
            this.mazo.inhabilitarCartaEspecial(this);
    }
    
    
    // Construye la sopresa que envia a la carcel
    Sorpresa(TipoSorpresa tipo, Tablero tablero){
        this.init();
        this.tablero = tablero;
        this.tipo    = tipo;
        
        this.valor   = tablero.getCarcel();
        this.texto   = "Envia a la carcel";
    }
    
    
    // Construye la sorpresa que envia a otra casilla
    Sorpresa(TipoSorpresa tipo, Tablero tablero, int valor, String texto){
        this.init();
        this.tablero = tablero;
        
        this.tipo    = tipo;
        
        this.valor   = valor;
        this.texto   = texto;
    }
   
    
    // Construye todas las sorpresas menos iracasilla, evitarcarcel e ircarcel
    Sorpresa(TipoSorpresa tipo, int valor, String texto){
        this.init();
        
        this.tipo  = tipo;
        
        this.valor = valor;
        this.texto = texto; 
    }
    
    
    
    // Construye la sorpresa que permite evitar la carcel
    Sorpresa(TipoSorpresa tipo, MazoSorpresas mazo){
        this.init();
        this.mazo = mazo;
        
        this.tipo = tipo;
        
        this.texto = "Evita la carcel";
    }
    
    
    /**
     * @brief Devuelve el nombre de la sorpresa(tipo)
     * @return Un string con el nombre
     */
    @Override
    public String toString(){
        return (    "TipoSorpresa: " + this.tipo.toString()
                  + "\nValor       " + this.valor
                  + "\nTexto:      " + this.texto
               );
    }
    
    
    /**
     * @brief Si la sorpresa es la que evita la carcél, habilita la
     *        carta especial en el mazo de sorpresas
     * @return void
     */
    void usada(){
        if(this.tipo == TipoSorpresa.SALIRCARCEL)
            mazo.habilitarCartaEspecial(this);
    }
    
    
    /**
     * @brief Metodo para probar la clase
     * @param args 
     *
    public static void main(String[] args){
        
        int casillaCarcel = 5;
        Tablero tablero = new Tablero(casillaCarcel);
        MazoSorpresas mazo = new MazoSorpresas();
        
        // Creamos varias sorpresas
        Sorpresa enviaCarcel  = new Sorpresa(TipoSorpresa.IRCASILLA, tablero);
        Sorpresa enviaCasilla = new Sorpresa(TipoSorpresa.IRCASILLA, 0, "Salida");
        Sorpresa beneficios = new Sorpresa(TipoSorpresa.PORJUGADOR, 50, "Fiesta");
        Sorpresa evitaCarcel  = new Sorpresa(TipoSorpresa.SALIRCARCEL, mazo);
        
        // Las mostramos
        System.out.println(enviaCarcel.toString() + "\n");
        System.out.println(enviaCasilla.toString()+ "\n");
        System.out.println(beneficios.toString()  + "\n");
        System.out.println(evitaCarcel.toString() + "\n");
    }
    */
        
} // Fin clase SORPRESA
